File: python/linked_list.py
from ast import Nonlocal
from locale import currency


class Node:
    def __init__(self, data, next=None):
        self.data = data 
        self.next = next 
        

class LinkedList:
    def __init__(self):
        self.head = None
        
    def insert_end(self, data):
        new_node = Node(data)
        if self.head != None:
            current_node = self.head
            while current_node.next:
                current_node = current_node.next
            current_node.next = new_node
        else:
            self.head = new_node

    def insert_start(self, data):
        new_node = Node(data)
        new_node.next = self.head 
        self.head = new_node

    def display(self):
        current_node = self.head 
        while current_node:
            print(current_node.data)
            current_node = current_node.next
            
    def insert_middle(self):
        size = self.get_size()
        if (size % 2 == 0):
            mid_point = size / 2
        else:
            mid_point = (size+1) / 2
        return mid_point
    
    def insert_after_item(self, item, data):
        new_node = Node(data)
        current_node = self.head 
        while current_node:
            if current_node.data == item:
                break
            current_node = current_node.next 
        if current_node is None:
            print("No item in the linked list")
        else:
            new_node.next = current_node.next 
            current_node.next = new_node
    
    def insert_before_an_item(self, item, data):
        #Checking if the head is empty
        if self.head is None: 
            print("List is empty")
        
        # Checking if the item is located in the first index
        if self.head.data == item:
            new_node = Node(data)
            new_node.next = self.head
            self.head = new_node 
        
        # Item does not exist in the first index
        current_node = self.head
        while current_node.next:
            if current_node.next.data == item:
                break
            current_node = current_node.next
        if current_node.next is None:
            print("Item not in the list")
        else:
            new_node = Node(data)
            new_node.next = current_node.next 
            current_node.next = new_node
            
    def insert_item_at_index(self, index, data):
        # If index is at position 1
        if index > 0:
            new_node = Node(data)
            if index == 1:
                new_node.next = self.head
                self.head = new_node 
            else:
                # Indext not at position 1
                # Loop through the list and insert node at the required index 
                current_node = self.head 
                i = 1
                while i < index-1 and current_node is not None:
                    current_node = current_node.next 
                    i+=1
                if current_node is None:
                    print("Index out of range")
                else:
                    new_node = Node(data)
                    new_node.next = current_node.next 
                    current_node.next = new_node
        else:
            print(index, "Not valid index, index must be greater than 0")
        return  
            
    def search(self, data):
        current_node = self.head 
        while current_node:
            if data == current_node.data:
                return f"Node found {current_node.data}"
            current_node = current_node.next
        return "Not found"
    
    def get_size(self):
        current_node = self.head
        count = 0
        while current_node:
            count+= 1
            current_node = current_node.next 
        return count   
    
    def delete_item_by_value(self, item):
        if self.head is None:
            print("List is epmty")
            return
        # if item is in the first node 
        if self.head.data == item:
            self.head = self.head.next
            return
        # if item does not exist in the first node
        current_node = self.head 
        while current_node.next:
            if current_node.next.data == item:
                break
            current_node = current_node.next
             
        if current_node.next is None:
            print("Item does not exist")
        else:
            current_node.next = current_node.next.next 
            
            
    def reverse(self):
        previous_node = None 
        current_node = self.head 
        while current_node:
            next_node = current_node.next
            current_node.next = previous_node
            previous_node = current_node
            current_node = next_node
        self.head = previous_node    
            
    def delete_end(self):
        current_node = self.head 
        while current_node.next.next != None:
            current_node = current_node.next
        current_node.next = None
        
    def delete_start(self):
        self.head = self.head.next
            
print('----++++++------')
llst = LinkedList()
llst.insert_start(60)
llst.insert_start(80)
llst.display()
print('----++++++------')
llst.insert_end(5)
llst.insert_end(3)
llst.display()
print('----++++++------')
llst.insert_after_item(5, 450)
llst.display()
print('----++++++------')
llst.insert_before_an_item(5, 333)
llst.insert_item_at_index(1, 870)
llst.display()
print('----++++++------')
llst.delete_item_by_value(870)
llst.display()
print('----++++++------')
llst.reverse()
llst.display()
print('----++++++------')





                 



                 
                 
                 
                 
                 

File: javascript/linkedList/LinkedList.js
class Node{
	constructor(data, next=null){
		this.data = data;
		this.next = next;
	}
}

class LinkedList{
	constructor(){
		this.head = null;
	}

	insertAtTheBegining(data){
		let newNode = new Node(data)
		newNode.next = this.head;
		this.head = newNode;
	}

	insertAtTheEnd(data){
		let newNode = new Node(data)
		if(this.head){
			let currentNode = this.head
			while(currentNode.next){
				currentNode = currentNode.next
			}
			currentNode.next = newNode;
		}else{
			this.head = newNode;
		}
	}

    insertItemAfter(item, data){
			let newNode = new Node(data);
			let currentNode = this.head;
			while(currentNode){
				if(currentNode.data === item){
					break
				}
			currentNode = currentNode.next;
			}
			if(currentNode == null){
				console.log("No item in the list")
			}else{
				newNode.next = currentNode.next;
				currentNode.next = newNode;
			}
    }

    insertItemBefore(item, data){
			if(this.head == null){
				console.log("The list is empty")
			}
			if(this.head.data == item){
				let newNode = new Node(data)
				newNode.next = this.head;
				this.head = newNode;
			}
			let currentNode = this.head;
			while(currentNode.next){
				if(currentNode.next.data == item){
					break
				}
			currentNode = currentNode.next
			}
			if(currentNode.next == null){
				console.log("Item not in the list")
			}else{
				let newNode = new Node(data)
				newNode.next = currentNode.next;
				currentNode.next = newNode;
			}
    }

    insertAtIndex(index, data){
			if(index > 0){
				if(index == 1){
					let newNode = new Node(data);
					newNode.next = this.head;
					this.head = newNode;
				}else{
					let currentNode = this.head;
					let i = 1;
					while(i < index - 1 && currentNode != null){
						i+=1;
						currentNode = currentNode.next;
					}
					if(currentNode == null){
						console.log("Index out of range");
					}else{
						let newNode = new Node(data);
						newNode.next = currentNode.next;
						currentNode.next = newNode;
					}
				}
		}else{
			console.log(`${index} is invalid: index must be greator than zero`)
		}
	}
    
    getSize(){
			let count = 0;
			let currentNode = this.head;
			while(currentNode){
				count+= 1;
				currentNode = currentNode.next;
			}
			return count;
    }

    deleteItemAtStart(){
			if(this.head == null){
				console.log("The list is empty");
			}
			this.head = this.head.next;
    }

    deleteItemAtTheEnd(){
			if(this.head == null){
				console.log("The list is empty");
			}
			let currentNode = this.head;
			while(currentNode.next.next){
				currentNode = currentNode.next;
			}
			currentNode.next = null;
    }

    deleteItemByValue(item){
			// check if list is empty
			if(this.head == null){
				console.log("List is empty")
			}
			// check if the item is in the first Node 
			if(this.head.data === item){
				this.head = this.head.next;
			}
			// Check if item exists in susbsequent Nodes 
			let currentNode = this.head;
			while(currentNode.next){
				if(currentNode.next.data == item){
					break
				}
				currentNode = currentNode.next;
			}
			if(currentNode.next == null){
				console.log("Item does not exist")
			}else{
				currentNode.next = currentNode.next.next;
			}
    }
    display(){
			let currentHead = this.head
			while(currentHead){
				console.log(currentHead.data)
				currentHead = currentHead.next
			}
    }

    search(data){
			let currentNode = this.head; 
			while(currentNode){
				if(data === currentNode.data){
					return currentNode.data;
				}
				currentNode = currentNode.next;
			}
			return "Node not found";
    }

    reverseLinkedList(){
			let currentNode = this.head;
			let previousNode = null;

			while(currentNode != null){
				let nextNode = currentNode.next;
				currentNode.next = previousNode;
				previousNode = currentNode;
				currentNode = nextNode;
			}
			this.head = previousNode;
	}

    bubleSortDataExchange(){
			let end = null;
			while(end != this.head){
				let current_node = this.head;
				while(current_node.next != end){
					let q_node = current_node.next;
					if(current_node.data > q_node.data){
						[current_node.data, q_node.data] = [q_node.data, current_node.data];
					}
					current_node = current_node.next
				} 
				end = current_node;
			} 
	}
}


let one = new LinkedList();
one.insertAtTheBegining(60);
one.insertAtTheBegining(50);
one.insertAtTheEnd(5);
one.insertAtTheEnd(4);
one.insertItemAfter(60, 500)
one.insertItemBefore(5, 888);
one.bubleSortDataExchange();
one.display()





